Storage/container/one2one.h

```diff
@@ -0,0 +1,253 @@
+      }
+    }
+    journal_.LogMutation(T_UPDATE_EVENT(object), [this, row_col]() { DoErase(row_col); });
+    DoAdd(row_col, object);
+  }
+
+  void Erase(const T_RCPAIR& row_col) {
+    auto it = map_.find(row_col);
+    if (it != map_.end()) {
+      const T prev_object = *(it->second);
+      journal_.LogMutation(T_DELETE_EVENT(prev_object),
+                           [this, row_col, prev_object]() { DoAdd(row_col, prev_object); });
+      DoErase(row_col);
+    }
+  }
+  void Erase(sfinae::CF<T_ROW> row, sfinae::CF<T_COL> col) { Erase(std::make_pair(row, col)); }
+
+  void EraseRow(sfinae::CF<T_ROW> row) {
+    const auto it = forward_.find(row);
+    if (it != forward_.end()) {
+      const T prev_object = *(it->second);
+      const auto row_col = std::make_pair(row, sfinae::GetCol(prev_object));
+      journal_.LogMutation(T_DELETE_EVENT(prev_object),
+                           [this, row_col, prev_object]() { DoAdd(row_col, prev_object); });
+      DoErase(row_col);
+    }
+  }
+
+  void EraseCol(sfinae::CF<T_COL> col) {
+    const auto it = transposed_.find(col);
+    if (it != transposed_.end()) {
+      const T prev_object = *(it->second);
+      const auto row_col = std::make_pair(sfinae::GetRow(prev_object), col);
+      journal_.LogMutation(T_DELETE_EVENT(prev_object),
+                           [this, row_col, prev_object]() { DoAdd(row_col, prev_object); });
+      DoErase(row_col);
+    }
+  }
+
+  ImmutableOptional<T> operator[](const T_RCPAIR& row_col) const {
+    const auto it = map_.find(row_col);
+    if (it == map_.end()) return nullptr;
+    return ImmutableOptional<T>(FromBarePointer(), it->second.get());
+  }
+  ImmutableOptional<T> Get(sfinae::CF<T_ROW> row, sfinae::CF<T_COL> col) const {
+    return operator[](std::make_pair(row, col));
+  }
+  ImmutableOptional<T> GetRow(sfinae::CF<T_ROW> row) const {
+    const auto it = forward_.find(row);
+    if (it == forward_.end()) return nullptr;
+    return ImmutableOptional<T>(FromBarePointer(), it->second);
+  }
+  ImmutableOptional<T> GetCol(sfinae::CF<T_COL> col) const {
+    const auto it = transposed_.find(col);
+    if (it == transposed_.end()) return nullptr;
+    return ImmutableOptional<T>(FromBarePointer(), it->second);
+  }
+
+  bool CanAdd(const T_RCPAIR& row_col) const {
+    return map_.find(row_col) == map_.end() && forward_.find(row_col.first) == forward_.end() &&
+           transposed_.find(row_col.second) == transposed_.end();
+  }
+  bool CanAdd(sfinae::CF<T_ROW> row, sfinae::CF<T_COL> col) const { return CanAdd(std::make_pair(row, col)); }
+
+  void operator()(const T_UPDATE_EVENT& e) {
+    const auto row = sfinae::GetRow(e.data);
+    const auto col = sfinae::GetCol(e.data);
+    DoAdd(std::make_pair(row, col), e.data);
+  }
+  void operator()(const T_DELETE_EVENT& e) { DoErase(std::make_pair(e.key.first, e.key.second)); }
+
+  template <typename T_MAP>
+  struct MapAccessor final {
+    using T_KEY = typename T_MAP::key_type;
```

Comment:
@mzhurovich , do you think it's a good time to start migrating to `key_t`, and also add top-level template types as `using map_t = T_MAP` ?